service/frontend/http_api_server.go

```diff
@@ -0,0 +1,438 @@
+	}
+
+	// Build 4 possible marshalers in order based on content type
+	opts := []runtime.ServeMuxOption{
+		runtime.WithMarshalerOption("application/json+pretty+no-payload-shorthand", h.newMarshaler(true, true)),
+		runtime.WithMarshalerOption("application/json+no-payload-shorthand", h.newMarshaler(false, true)),
+		runtime.WithMarshalerOption("application/json+pretty", h.newMarshaler(true, false)),
+		runtime.WithMarshalerOption(runtime.MIMEWildcard, h.newMarshaler(false, false)),
+	}
+
+	// Set Temporal service error handler
+	opts = append(opts, runtime.WithProtoErrorHandler(h.errorHandler))
+
+	// Match headers w/ default
+	h.matchAdditionalHeaders = map[string]bool{}
+	for _, v := range defaultForwardedHeaders {
+		h.matchAdditionalHeaders[v] = true
+	}
+	for _, v := range rpcConfig.HTTPAdditionalForwardedHeaders {
+		h.matchAdditionalHeaders[http.CanonicalHeaderKey(v)] = true
+	}
+	opts = append(opts, runtime.WithIncomingHeaderMatcher(h.incomingHeaderMatcher))
+
+	// Create inline client connection
+	clientConn := newInlineClientConn(
+		map[string]any{"temporal.api.workflowservice.v1.WorkflowService": handler},
+		interceptors,
+	)
+
+	// Create serve mux
+	h.serveMux = runtime.NewServeMux(opts...)
+	err = workflowservice.RegisterWorkflowServiceHandlerClient(
+		context.Background(),
+		h.serveMux,
+		workflowservice.NewWorkflowServiceClient(clientConn),
+	)
+	if err != nil {
+		return nil, fmt.Errorf("failed registering HTTP API handler: %w", err)
+	}
+	// Set the handler as our function that wraps serve mux
+	h.server.Handler = http.HandlerFunc(h.serveHTTP)
+
+	// Put the remote address on the context
+	h.server.ConnContext = func(ctx context.Context, c net.Conn) context.Context {
+		return context.WithValue(ctx, httpRemoteAddrContextKey{}, c)
+	}
+
+	// We want to set ReadTimeout and WriteTimeout as max idle (and IdleTimeout
+	// defaults to ReadTimeout) to ensure that a connection cannot hang over that
+	// amount of time.
+	h.server.ReadTimeout = serviceConfig.KeepAliveMaxConnectionIdle()
+	h.server.WriteTimeout = serviceConfig.KeepAliveMaxConnectionIdle()
+
+	success = true
+	return h, nil
+}
+
+// Serve serves the HTTP API and does not return until there is a serve error or
+// GracefulStop completes. Upon graceful stop, this will return nil. If an error
+// is returned, the message is clear that it came from the HTTP API server.
+func (h *HTTPAPIServer) Serve() error {
+	err := h.server.Serve(h.listener)
+	// If the error is for close, we have to wait for the shutdown to complete and
+	// we don't consider it an error
+	if err == http.ErrServerClosed {
+		<-h.stopped
+		err = nil
+	}
+	// Wrap the error to be clearer it's from the HTTP API
+	if err != nil {
+		return fmt.Errorf("HTTP API serve failed: %w", err)
+	}
+	return nil
+}
+
+// GracefulStop stops the HTTP server. This will first attempt a graceful stop
+// with a drain time, then will hard-stop. This will not return until stopped.
+func (h *HTTPAPIServer) GracefulStop() {
```

Comment:
This should accept a context because we have one at the top-level from fx (although you may have to plumb it down from there)

Reply:
We don't use the context for bounding gRPC server graceful stop, why is it a new requirement for HTTP server graceful stop? I intentionally chose to match what gRPC server does.